front_end/audits2/Audits2Panel.js

```diff
@@ -1,43 +1,242 @@
+    }
+    this._start();
+  }
+
+  _updateButton() {
+    this._startButton.textContent = this._auditRunning ? Common.UIString('Cancel audit') : Common.UIString('Audit this page');
+    this._startButton.classList.toggle('started', this._auditRunning);
+    this._statusView.classList.toggle('hidden', !this._auditRunning);
+  }
+
+  /**
+   * @param {string} statusMessage
+   */
+  _updateStatus(statusMessage) {
+    this._statusElement.textContent = Common.UIString(statusMessage);
+  }
+
+  _stop() {
+    this._protocolService.detach().then(_ => {
+      this._auditRunning = false;
+      this._updateButton();
+      if (this._inspectedURL !== SDK.targetManager.mainTarget().inspectedURL())
+          SDK.targetManager.mainTarget().pageAgent().navigate(this._inspectedURL);
+
     });
   }
 
   /**
+   * @param {!Audits2.WorkerResult} result
+   */
+  _finish(result) {
+    this._stop();
+    this.resultsView.removeChildren();
+
+    this.resultsView.appendChild(this._createResultsBar(result.lighthouseResult.url, result.lighthouseResult.generatedTime));
+    this.resultsView.appendChild(this._createIframe(result.bloburl));
+    this.contentElement.classList.add('show-results');
+  }
+
+  /**
+   * @param {string} bloburl
+   * @return {!Element}
+   */
+  _createIframe(bloburl) {
+    var iframeContainer = createElementWithClass('div', 'iframe-container');
+    var iframe = iframeContainer.createChild('iframe', 'fill');
+    iframe.setAttribute('sandbox', 'allow-scripts allow-popups-to-escape-sandbox allow-popups');
+    iframe.src = bloburl;
+    return iframeContainer;
+  }
+
+  /**
+   * @param {string} url
+   * @param {string} timestamp
+   * @return {!Element}
+   */
+  _createResultsBar(url, timestamp) {
+    var elem = createElementWithClass('div', 'results-bar hbox');
+    elem.createChild('div', 'logo-small');
+
+    var summaryElem = elem.createChild('div', 'summary');
+    var reportFor = summaryElem.createChild('span');
+    reportFor.innerHTML = `Report for <b>${url}</b>`;
+    var timeElem = summaryElem.createChild('span');
+    timeElem.textContent =
+        `Generated at ${new Date(timestamp).toLocaleDateString()} ${new Date(timestamp).toLocaleTimeString()}`;
+
+    var newAuditButton = createTextButton(Common.UIString('New Audit'), this._reset.bind(this), 'new-audit audit-btn');
+    elem.appendChild(newAuditButton);
+    return elem;
+  }
+};
+
+Audits2.ProtocolService = class extends Common.Object {
+  constructor() {
+    super();
+    this._rawConnection = undefined;
+    this._backend = undefined;
+    this._backendPromise = undefined;
+    this._status = undefined;
+  }
+
+  /**
+   * @return {!Promise<undefined>}
+   */
+  attach() {
+    return SDK.targetManager.interceptMainConnection(this._dispatchProtocolMessage.bind(this))
+        .then(rawConnection => {
+          this._rawConnection = rawConnection;
+        });
+  }
+
+  /**
+   * @param {string} inspectedURL
+   * @return {!Promise<!Object|undefined>}
```

Comment:
based on `_finish`'s jsdoc, does this actually return `!Promise<!Audits2.WorkerResult>`? (you might need to cast internally since `this._send` *does* resolve on the generic `!Object|undefined`)